### PR TITLE
feat: implement authenticateRequest and verifyToken helper methods

### PR DESCRIPTION
**Pull request by 2ynn**
This PR implements [authenticateRequest](https://clerk.com/docs/references/backend/authenticate-request) and [verifyToken](https://clerk.com/docs/references/backend/verify-token) helper methods.

#### Notes
- the `src/main/java/com/clerk/backend_api/helpers/` folder is *.genignored*
- unlike in the `clerk-sdk-python` implementation, the `secretKey` should be passed manually. Due to [sdkConfiguration](https://github.com/clerk/clerk-sdk-java/blob/a5124d4f5a6bd696163469fa26dc487b726214dd/src/main/java/com/clerk/backend_api/Clerk.java#L194) being a private member of `Clerk` class, the `bearerAuth` value passed during sdk instantiation cannot be reused as `secretKey` for convenience. 

#### Limitations
- the added helper functions are only applicable for Backend APIs, `afterSignInUrl`/`afterSignUpUrl` options are not implemented
- multi-domain setup (`isSatellite`, `proxyUrl`, `signInUrl`, `signUpUrl`) is not implemented
- caching is not covered by this PR and `skipJwksCache` option is not made available

#### Tests
To run tests (`./gradlew test`) the following environment variables should be set:
- `CLERK_SESSION_TOKEN`: The session token to be tested.
- `CLERK_SECRET_KEY`: The Clerk secret key from the API Keys page in the Clerk Dashboard (needed for fetching Clerk's Jwks)
- `CLERK_JWT_KEY`: The PEM public key from Clerk Dashboard (needed for networkless verification only)

#### Example Usage
1. Remote JWKS (using `secretKey`)
```java
import java.net.http.HttpRequest;
import com.clerk.backend_api.helpers.jwks.AuthenticateRequest;
import com.clerk.backend_api.helpers.jwks.AuthenticateRequestOptions;
import com.clerk.backend_api.helpers.jwks.RequestState;

public class UserAuthentication {

    public static boolean isSignedIn(HttpRequest request) {
        RequestState state = AuthenticateRequest.authenticateRequest(request, AuthenticateRequestOptions
                .secretKey(System.getenv("CLERK_SECRET_KEY"))
                .authorizedParty("https://example.com")
                .build());
        return state.isSignedIn();
    }
```

2. Networkless (using local PEM formatted `jwtKey`)
```java
import java.net.http.HttpRequest;
import com.clerk.backend_api.helpers.jwks.AuthenticateRequest;
import com.clerk.backend_api.helpers.jwks.AuthenticateRequestOptions;
import com.clerk.backend_api.helpers.jwks.RequestState;

public class UserAuthentication {

   public static boolean isSignedIn(HttpRequest request) {
        RequestState state = AuthenticateRequest.authenticateRequest(request, AuthenticateRequestOptions
                .jwtKey(System.getenv("CLERK_JWT_KEY"))
                .authorizedParty("https://example.com")
                .build());
        return state.isSignedIn();
    }
}
```

### CODE REVIEW

File: src/main/java/com/clerk/backend_api/helpers/jwks/TokenVerificationException.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.clerk.backend_api.helpers.jwks;
+
+/**
+ * TokenVerificationException - Exception thrown when token verification fails.
+ */
+@SuppressWarnings("serial")
+public final class TokenVerificationException extends RuntimeException {
```
Comment (alex-ntousias) [inline src/main/java/com/clerk/backend_api/helpers/jwks/TokenVerificationException.java:7]: It's probably better to have the `TokenVerificationException` as a checked exception and not `RuntimeException`, to force users to handle it.

**Review by alex-ntousias — COMMENTED** [review]
Overall, it looks good to me! I've added a few comments to be addressed.
Just to confirm, you've tested it and everything works as expected, right?

File: src/main/java/com/clerk/backend_api/helpers/jwks/AuthenticateRequestOptions.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    public long clockSkewInMs() {
+        return clockSkewInMs;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static BuilderWithKey secretKey(String secretKey) {
```
[the remark this answers is not in the document]
Comment (2ynn) [reply src/main/java/com/clerk/backend_api/helpers/jwks/AuthenticateRequestOptions.java]: These two methods are now made available:
```
    public static Builder secretKey(String secretKey) {
        return Builder.withSecretKey(secretKey);
    }

    public static Builder jwtKey(String jwtKey) {
        return Builder.withJwtKey(jwtKey);
    }
```
let me know if I missed the mark

File: src/main/java/com/clerk/backend_api/helpers/jwks/VerifyToken.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+     *                the public API (subject to change).
+     */
+    public static Claims verifyToken(String token, VerifyTokenOptions options) {
+
+        Key key;
+        if (options.jwtKey().isPresent()) {
+            key = getLocalJwtKey(options.jwtKey().get());
+        } else {
+            key = getRemoteJwtKey(token, options);
+        }
```
[the remark this answers is not in the document]
Comment (2ynn) [reply src/main/java/com/clerk/backend_api/helpers/jwks/VerifyToken.java:58]: after changing `TokenVerificationException` to a checked exception it cannot be thrown directly from within `map` and `orElseGet` lambda expressions,  which means I would need to add additional try/catch blocks and wrap caught exceptions into `RuntimeException`s.  I think leaving as is is probably more straightforward

**Comment by 2ynn** [comment]
> st to confirm, you've tested it and everything works as expected, right?

Thans

**Comment by 2ynn** [comment]
> Overall, it looks good to me! I've added a few comments to be addressed. Just to confirm, you've tested it and everything works as expected, right?

Thanks for the review @alex-ntousias ! 

To run tests locally I have:
1. created an account and grabbed the secretKey (`CLERK_SECRET_KEY`)+ PEM public key (`CLERK_JWT_KEY`) from the dashboard account
2. created a barebone nextjs app that I ran on `localhost:3000`
3. signed in using github
4. retrieved the session cookie from the browser  (`CLERK_SESSION_TOKEN`)
5. ran `./gradlew test` successfully
6. ran tests again: tests still pass but throw warnings to indicate the token is expired.

File: src/main/java/com/clerk/backend_api/helpers/jwks/AuthenticateRequestOptions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-ntousias) [reply src/main/java/com/clerk/backend_api/helpers/jwks/AuthenticateRequestOptions.java]: Looks good, thank you @2ynn 🙏

**Review by alex-ntousias — APPROVED** [review]
Nice work @2ynn , thank you for addressing my comments. 🚀

**Pull request merged**